neuralop/training/callbacks.py

```diff
@@ -375,82 +376,123 @@ def on_before_loss(self, out):
     
     def on_before_val_loss(self, **kwargs):
         return self.on_before_loss(**kwargs)
-
-class ModelCheckpointCallback(Callback):
-
-    def __init__(self, checkpoint_dir: Union[Path, str] = Path('./checkpoints'), interval: int = 1):
-        """
-        Implements basic model checkpointing by saving a model every N epochs.
         
+class StateCheckpointCallback(Callback):
+    
+    def __init__(self, 
+                 save_dir: Union[Path, str] = './checkpoints', 
+                 save_best : str = None,
+                 save_step : int = 1,
+                 save_optimizer : bool = False,
+                 save_scheduler : bool = False,
+                 save_regularizer : bool = False,
+                 resume_from_dir : Union[Path, str] = None
+                 ):
+        """StateCheckpointCallback handles saving and resuming 
+        training state from checkpoint .pt save files. 
+
         Parameters
         ----------
-        checkpoint_dir : str | pathlib.Path
-            folder in which to save checkpoints
-        interval : int
-            interval at which to check metric
+        save_dir : Union[Path, str], optional
+            folder in which to save checkpoints, by default './checkpoints'
+        save_best : str, optional
+            metric to monitor for best value in order to save state
+        save_step : int, optional
+            interval on which to save/check metric, by default 1
+        save_optimizer : bool, optional
+            whether to save optimizer state, by default False
+        save_scheduler : bool, optional
+            whether to save scheduler state, by default False
+        save_regularizer : bool, optional
+            whether to save regularizer state, by default False
+        resume_from_dir : Union[Path, str], optional
+            folder from which to resume training state. 
+            Expects saved states in the form: (all but model optional)
+                model.pt, optimizer.pt, scheduler.pt, regularizer.pt
+            All state files present will be loaded. 
```

Comment:
I'm wondering whether we should 
i) try to load best_model.pt (if monitor_metric is not None this will exist)
ii) if that fails, load model.pt (which will be the model from the latest save_interval)

Reply:
Currently isn't the way I have things set up - at each save_interval I create a new folder that saves the state. Same for best_model. Do you think it would be better to continually overwrite the save in one folder?

Reply:
I think so - to avoid blowing up memory?

Reply:
Done